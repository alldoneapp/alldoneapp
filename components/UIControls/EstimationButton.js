import React, { useRef, useState } from 'react'
import Popover from 'react-tiny-popover'
import Hotkeys from 'react-hot-keys'
import { useDispatch, useSelector } from 'react-redux'

import EstimationModal from '../UIComponents/FloatModals/EstimationModal/EstimationModal'
import Button from './Button'
import { hideFloatPopup, showFloatPopup } from '../../redux/actions'
import { execShortcutFn } from '../../utils/HelperFunctions'
import { translate } from '../../i18n/TranslationService'
import { ESTIMATION_0_MIN, getEstimationIconByValue } from '../../utils/EstimationHelper'
import { setTaskAutoEstimation } from '../../utils/backends/Tasks/tasksFirestore'
import { getTaskAutoEstimation, OPEN_STEP } from '../TaskListView/Utils/TasksHelper'

export default function EstimationButton({
    task,
    projectId,
    disabled,
    setEstimationFn,
    onDismissPopup,
    shortcutText,
    style,
    isObservedTask,
    isToReviewTask,
    editing,
    setTempAutoEstimation,
    isPending,
}) {
    const dispatch = useDispatch()
    const smallScreen = useSelector(state => state.smallScreen)
    const currentUserId = useSelector(state => state.currentUser.uid)
    const [visiblePopover, setVisiblePopover] = useState(false)

    let buttonRef = useRef().current

    const { estimations, stepHistory, estimationsByObserverIds } = task
    const currentStepId = stepHistory[stepHistory.length - 1]
    const currentEstimation = isPending
        ? estimations[OPEN_STEP]
        : isObservedTask && !isToReviewTask
        ? estimationsByObserverIds[currentUserId]
        : estimations[currentStepId] || ESTIMATION_0_MIN

    const hidePopover = () => {
        setTimeout(async () => {
            setVisiblePopover(false)
            dispatch(hideFloatPopup())

            if (onDismissPopup) onDismissPopup()
        })
    }

    const showPopover = () => {
        if (!visiblePopover) {
            setVisiblePopover(true)
            dispatch(showFloatPopup())
        }
    }

    const closePopover = () => {
        setVisiblePopover(false)
        hidePopover()
    }

    const setAutoEstimation = autoEstimation => {
        if (editing) setTaskAutoEstimation(projectId, task, autoEstimation)
        setTempAutoEstimation(autoEstimation)
    }

    return (
        <Popover
            content={
                <EstimationModal
                    projectId={projectId}
                    estimation={currentEstimation}
                    setEstimationFn={setEstimationFn}
                    closePopover={closePopover}
                    autoEstimation={getTaskAutoEstimation(projectId, currentEstimation, task.autoEstimation)}
                    setAutoEstimation={setAutoEstimation}
                    showAutoEstimation={!task.isSubtask}
                    disabled={!!task.calendarData}
                />
            }
            onClickOutside={hidePopover}
            isOpen={visiblePopover}
            position={['bottom', 'left', 'right', 'top']}
            padding={4}
            align={'end'}
            contentLocation={smallScreen ? null : undefined}
        >
            <Hotkeys
                keyName={`alt+${shortcutText}`}
                disabled={disabled}
                onKeyDown={(sht, event) => execShortcutFn(buttonRef, showPopover, event)}
                filter={e => true}
            >
                <Button
                    ref={ref => (buttonRef = ref)}
                    title={smallScreen ? null : translate('Estimation')}
                    type={'ghost'}
                    noBorder={smallScreen}
                    icon={`count-circle-${getEstimationIconByValue(projectId, currentEstimation)}`}
                    buttonStyle={style}
                    onPress={showPopover}
                    disabled={disabled}
                    shortcutText={shortcutText}
                />
            </Hotkeys>
        </Popover>
    )
}
