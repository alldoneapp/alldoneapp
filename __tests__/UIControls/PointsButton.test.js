/**
 * @jest-environment jsdom
 */

import React from 'react'
import EstimationButton from '../../components/UIControls/EstimationButton'

import renderer from 'react-test-renderer'

describe('Task Estimation Button component', () => {
    describe('Task Estimation Button snapshot test', () => {
        it('should render correctly', () => {
            const tree = renderer
                .create(<EstimationButton projectId={'-Asd'} task={{ id: '-Sda', name: 'My Task' }} />)
                .toJSON()
            expect(tree).toMatchSnapshot()
        })
    })

    describe('Function updateState snapshot test', () => {
        it('should execute and render correctly', () => {
            const tree = renderer.create(
                <EstimationButton projectId={'-Asd'} task={{ id: '-Sda', name: 'My Task' }} />
            )
            expect(tree.toJSON()).toMatchSnapshot()

            tree.getInstance().updateState()
            expect(tree.toJSON()).toMatchSnapshot()
        })
    })

    describe('Function hidePopover snapshot test', () => {
        it('should execute and render correctly', () => {
            const tree = renderer.create(
                <EstimationButton projectId={'-Asd'} task={{ id: '-Sda', name: 'My Task' }} />
            )
            expect(tree.toJSON()).toMatchSnapshot()

            tree.getInstance().state.visiblePopover = true
            tree.getInstance().hidePopover()
        })
    })

    describe('Task Estimation check unmount', () => {
        it('should unmount correctly', () => {
            const tree = renderer.create(
                <EstimationButton projectId={'-Asd'} task={{ id: '-Sda', name: 'My Task' }} />
            )
            tree.getInstance().componentWillUnmount()
        })
    })
})
