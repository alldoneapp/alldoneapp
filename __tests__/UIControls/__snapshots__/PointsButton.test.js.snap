// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task Points Button component Function hidePopover snapshot test should execute and render correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "flex-start",
      "backgroundColor": "transparent",
      "borderColor": "#C6CDD2",
      "borderRadius": 4,
      "borderStyle": "solid",
      "borderWidth": 1,
      "flexDirection": "row",
      "flexWrap": "nowrap",
      "height": 40,
      "justifyContent": "center",
      "maxHeight": 40,
      "minHeight": 40,
      "opacity": 1,
      "paddingHorizontal": 16,
      "paddingVertical": 8,
    }
  }
>
  <View>
    <Text
      style={
        Object {
          "color": "#8A94A6",
          "fontFamily": "alldone",
          "fontSize": 24,
          "opacity": 1,
        }
      }
    >
      
    </Text>
  </View>
  <Text
    numberOfLines={1}
    style={
      Array [
        Object {
          "alignSelf": "center",
          "color": "#FFFFFF",
          "flexWrap": "nowrap",
          "fontFamily": "Roboto-Medium",
          "fontSize": 14,
          "letterSpacing": 0.8,
          "lineHeight": 16,
          "margin": 0,
          "paddingHorizontal": 8,
          "paddingVertical": 0,
        },
        Object {
          "color": "#8A94A6",
        },
        Object {},
      ]
    }
  >
    Estimation
  </Text>
</View>
`;

exports[`Task Points Button component Function updateState snapshot test should execute and render correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "flex-start",
      "backgroundColor": "transparent",
      "borderColor": "#C6CDD2",
      "borderRadius": 4,
      "borderStyle": "solid",
      "borderWidth": 1,
      "flexDirection": "row",
      "flexWrap": "nowrap",
      "height": 40,
      "justifyContent": "center",
      "maxHeight": 40,
      "minHeight": 40,
      "opacity": 1,
      "paddingHorizontal": 16,
      "paddingVertical": 8,
    }
  }
>
  <View>
    <Text
      style={
        Object {
          "color": "#8A94A6",
          "fontFamily": "alldone",
          "fontSize": 24,
          "opacity": 1,
        }
      }
    >
      
    </Text>
  </View>
  <Text
    numberOfLines={1}
    style={
      Array [
        Object {
          "alignSelf": "center",
          "color": "#FFFFFF",
          "flexWrap": "nowrap",
          "fontFamily": "Roboto-Medium",
          "fontSize": 14,
          "letterSpacing": 0.8,
          "lineHeight": 16,
          "margin": 0,
          "paddingHorizontal": 8,
          "paddingVertical": 0,
        },
        Object {
          "color": "#8A94A6",
        },
        Object {},
      ]
    }
  >
    Estimation
  </Text>
</View>
`;

exports[`Task Points Button component Function updateState snapshot test should execute and render correctly 2`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "flex-start",
      "backgroundColor": "transparent",
      "borderColor": "#C6CDD2",
      "borderRadius": 4,
      "borderStyle": "solid",
      "borderWidth": 1,
      "flexDirection": "row",
      "flexWrap": "nowrap",
      "height": 40,
      "justifyContent": "center",
      "maxHeight": 40,
      "minHeight": 40,
      "opacity": 1,
      "paddingHorizontal": 16,
      "paddingVertical": 8,
    }
  }
>
  <View>
    <Text
      style={
        Object {
          "color": "#8A94A6",
          "fontFamily": "alldone",
          "fontSize": 24,
          "opacity": 1,
        }
      }
    >
      
    </Text>
  </View>
  <Text
    numberOfLines={1}
    style={
      Array [
        Object {
          "alignSelf": "center",
          "color": "#FFFFFF",
          "flexWrap": "nowrap",
          "fontFamily": "Roboto-Medium",
          "fontSize": 14,
          "letterSpacing": 0.8,
          "lineHeight": 16,
          "margin": 0,
          "paddingHorizontal": 8,
          "paddingVertical": 0,
        },
        Object {
          "color": "#8A94A6",
        },
        Object {},
      ]
    }
  >
    Estimation
  </Text>
</View>
`;

exports[`Task Points Button component Task Points Button snapshot test should render correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "flex-start",
      "backgroundColor": "transparent",
      "borderColor": "#C6CDD2",
      "borderRadius": 4,
      "borderStyle": "solid",
      "borderWidth": 1,
      "flexDirection": "row",
      "flexWrap": "nowrap",
      "height": 40,
      "justifyContent": "center",
      "maxHeight": 40,
      "minHeight": 40,
      "opacity": 1,
      "paddingHorizontal": 16,
      "paddingVertical": 8,
    }
  }
>
  <View>
    <Text
      style={
        Object {
          "color": "#8A94A6",
          "fontFamily": "alldone",
          "fontSize": 24,
          "opacity": 1,
        }
      }
    >
      
    </Text>
  </View>
  <Text
    numberOfLines={1}
    style={
      Array [
        Object {
          "alignSelf": "center",
          "color": "#FFFFFF",
          "flexWrap": "nowrap",
          "fontFamily": "Roboto-Medium",
          "fontSize": 14,
          "letterSpacing": 0.8,
          "lineHeight": 16,
          "margin": 0,
          "paddingHorizontal": 8,
          "paddingVertical": 0,
        },
        Object {
          "color": "#8A94A6",
        },
        Object {},
      ]
    }
  >
    Estimation
  </Text>
</View>
`;
